backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/timeout/impl/LuceneTimeoutManager.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.lucene.search.timeout.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+import org.hibernate.search.backend.lucene.logging.impl.Log;
+import org.hibernate.search.engine.common.timing.impl.TimingSource;
+import org.hibernate.search.engine.search.timeout.spi.TimeoutManager;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.Counter;
+
+/**
+ * @author Emmanuel Bernard
+ */
+public final class LuceneTimeoutManager extends TimeoutManager {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	public static LuceneTimeoutManager noTimeout(TimingSource timingSource, Query query) {
+		return new LuceneTimeoutManager( timingSource, query, null, null, Type.NONE );
+	}
+
+	public static LuceneTimeoutManager softTimeout(TimingSource timingSource, Query query, long timeout, TimeUnit timeUnit) {
+		return new LuceneTimeoutManager( timingSource, query, timeout, timeUnit, Type.LIMIT );
+	}
+
+	public static LuceneTimeoutManager hardTimeout(TimingSource timingSource, Query query, long timeout, TimeUnit timeUnit) {
+		return new LuceneTimeoutManager( timingSource, query, timeout, timeUnit, Type.EXCEPTION );
+	}
+
+	private enum Type {
+		NONE,
+		EXCEPTION,
+		LIMIT;
+	}
+
+	private final Query query;
+	private final Long timeoutValue;
+	private final TimeUnit timeoutUnit;
+	private final Type type;
+
+	private LuceneTimeoutManager(TimingSource timingSource, Query query, Long timeoutValue, TimeUnit timeoutUnit, Type type) {
+		super( timingSource, timeoutUnit == null ? null : timeoutUnit.toMillis( timeoutValue ) );
+		this.query = query;
+		this.timeoutValue = timeoutValue;
+		this.timeoutUnit = timeoutUnit;
+		this.type = type;
+
+		timingSource.ensureInitialized();
+	}
+
+	public Counter createCounter() {
+		return new LuceneCounterAdapter( timingSource );
+	}
+
+	@Override
+	protected void onTimedOut() {
+		if ( hasHardTimeout() ) {
```

Comment:
I think this should be moved to the superclass, too. Be it only because it's simpler that way, or because we also support "soft" timeouts with the Elasticsearch backend. You could get rid of a few class members in the Elasticsearch backend if it was able to query `hasHardTimeout` directly.

Reply:
Sure. Originally I didn't think that Elasicsearch could use it.